engine/src/main/java/org/pentaho/di/trans/steps/mergejoin/MergeJoin.java

```diff
@@ -0,0 +1,466 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.mergejoin;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleStepException;
+import org.pentaho.di.core.row.RowDataUtil;
+import org.pentaho.di.core.row.RowMeta;
+import org.pentaho.di.core.row.RowMetaInterface;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.BaseStep;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.step.StepMetaInterface;
+import org.pentaho.di.trans.step.errorhandling.StreamInterface;
+
+/**
+ * Merge rows from 2 sorted streams and output joined rows with matched key fields. Use this instead of hash join is
+ * both your input streams are too big to fit in memory. Note that both the inputs must be sorted on the join key.
+ *
+ * This is a first prototype implementation that only handles two streams and inner join. It also always outputs all
+ * values from both streams. Ideally, we should: 1) Support any number of incoming streams 2) Allow user to choose the
+ * join type (inner, outer) for each stream 3) Allow user to choose which fields to push to next step 4) Have multiple
+ * output ports as follows: a) Containing matched records b) Unmatched records for each input port 5) Support incoming
+ * rows to be sorted either on ascending or descending order. The currently implementation only supports ascending
+ *
+ * @author Biswapesh
+ * @since 24-nov-2006
+ */
+
+public class MergeJoin extends BaseStep implements StepInterface {
+  private static Class<?> PKG = MergeJoinMeta.class; // for i18n purposes, needed by Translator2!!
+
+  private MergeJoinMeta meta;
+  private MergeJoinData data;
+
+  public MergeJoin( StepMeta stepMeta, StepDataInterface stepDataInterface, int copyNr, TransMeta transMeta,
+    Trans trans ) {
+    super( stepMeta, stepDataInterface, copyNr, transMeta, trans );
+  }
+
+  public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws KettleException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 68. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.